### ISSUE TITLE
Which behavior is correct?

### ISSUE THREAD

**Issue by hax** [promises-aplus/promises-spec#183]
See https://github.com/hax/test-promises/blob/master/promises-spec.md

As my understanding of latest es6 draft, Firefox's implementation or my-promise implementation seem correct (my-promise seems more correct which promise job queue should prior to settimeout queue but I'm not sure about that).

But maybe some order may be not important. Especially the order of "a1 a2 a3 a4 a5 a6 b1 b2 b3 b4 b5 b6". It's relate to https://github.com/tj/co/issues/180#issuecomment-71443729 . When I write my-promise, I found that it's impossible to avoid the leak completely, each nest promise need at least one pointer to preserve the order. So if https://github.com/promises-aplus/promises-spec/issues/179 is a case should be considered, we need relax ES6 spec to allow different order.

**Comment by domenic** [comment]
I'll look into this more later but it's worth noting that in general Firefox is very bad at following the ES2015 spec, especially around ordering.

**Comment by hax** [comment]
I wrote the tests and my-promise implementation as the previous draft (rev 30). Are there any changes about promise in rev 31?

**Comment by petkaantonov** [comment]
It's not possible to avoid the leak at all if you implement ES6, there is no state adoption relaxation like there is in Promises/A+ - you must always call `x.then(...)`

**Comment by hax** [comment]
@domenic you are right, FF is not correct (I don't know why I thought it's correct, may be I slept too little :-)

@petkaantonov I think opt is possible, you just need to make sure it's instanceof Promise and the then method is same as PromisePrototype.then

**Comment by petkaantonov** [comment]
@hax That doesn't help though, the spec needs to be relaxed or a leak is forced (even small leak that you have is a problem for servers)

**Comment by hax** [comment]
@petkaantonov About "then" logs,  you may have noticed an interesting thing is that Chrome canary and IE 11 called "then" several times , which no other Promise/A+ implementations do. It seems be able to avoid leak also. I will test them after my vacation.

**Comment by petkaantonov** [comment]
`Version 42.0.2292.0 canary` still leaks as much as any other... there are no changes to V8's promise implementation in the last 2 months either.

**Comment by hax** [comment]
Any updates? ping @domenic  :)

**Comment by hax** [comment]
My update: Firefox 37 now have the same order result as my-promise which I believe is strictly follow the ES6 spec.
